rasa/core/featurizers.py

```diff
@@ -259,6 +287,314 @@ def create_encoded_all_actions(self, domain: Domain) -> np.ndarray:
+            slot_values = np.hstack(slot_values)
+            binary_features[: len(self.slot_states)] = slot_values
+        if state.get(FORM):
+            # featurize forms
+            form_values = np.zeros((len(self.form_states)))
+            form_values[self.form_states.index(state.get(FORM).get("name"))] += 1
+            binary_features[
+                len(self.slot_states) : len(self.slot_states + self.form_states)
+            ] = form_values
+        if state.get(USER):
+            if state[USER].get(ENTITIES):
+                entities = state[USER].get(ENTITIES)
+                for entity in entities:
+                    binary_features[
+                        len(self.slot_states + self.form_states)
+                        + self.entities.index(entity)
+                    ] += 1
+
+        return binary_features
+
+    def _check_dense_features(
+        self, dense_features: Tuple[np.ndarray, np.ndarray]
+    ) -> bool:
+        if dense_features[1] is None:
+            return False
+        if dense_features[1].size == 0:
+            return False
+        return True
+
+    def _record_shape(
+        self,
+        attribute: Text,
+        sparse_features: Tuple[scipy.sparse.spmatrix, scipy.sparse.spmatrix],
+        dense_features: Tuple[np.ndarray, np.ndarray],
+    ) -> None:
+        if attribute in DENSE_FEATURIZABLE_ATTRIBUTES:
+            if not sparse_features[1] is None:
+                if f"{attribute}_sparse" not in self.output_shapes.keys():
+                    self.output_shapes[f"{attribute}_sparse"] = sparse_features[
+                        1
+                    ].shape[-1]
+
+            if self._check_dense_features(dense_features):
+                if f"{attribute}_dense" not in self.output_shapes.keys():
+                    self.output_shapes[f"{attribute}_dense"] = dense_features[1].shape[
+                        -1
+                    ]
+        else:
+            if not sparse_features[0] is None:
+                if f"{attribute}" not in self.output_shapes.keys():
+                    self.output_shapes[f"{attribute}"] = sparse_features[0].shape[-1]
+
+    def _construct_message(self, state: Dict[Text, Text], state_comes_from: Text):
+        if state_comes_from == USER:
+            if state.get(INTENT):
+                message = Message(data={INTENT: state.get(INTENT)})
+                attribute = INTENT
+            else:
+                message = Message(state.get(TEXT))
+                attribute = TEXT
+        else:
+            if state.get(ACTION_NAME):
+                message = Message(data={ACTION_NAME: state.get(ACTION_NAME)})
+                attribute = ACTION_NAME
+            else:
+                message = Message(data={ACTION_TEXT: state.get(ACTION_TEXT)})
+                attribute = ACTION_TEXT
+        return message, attribute
+
+    def _extract_features(
+        self,
+        state: Optional[Dict[str, Union[str, Tuple[Union[float, str]]]]],
```

Comment:
is it of the type `STATE` I introduced in domain?

Reply:
No, it has one less `Dict` on the outside; it is either `state[USER]` or `state[PREV_ACTION]`.